conda/core/prefix_data.py

```diff
@@ -97,6 +142,80 @@ def __eq__(self, other: Any) -> bool:
             # neither prefix exists, raw comparison
             return self.prefix_path.resolve() == other.prefix_path.resolve()
 
+    def exists(self):
+        try:
+            return self.prefix_path.is_dir()
+        except OSError:
+            return False
+
+    def is_environment(self):
+        try:
+            return self._magic_file.is_file()
+        except OSError:
+            return False
+
+    def is_base(self):
+        return paths_equal(str(self.prefix_path), context.root_prefix)
+
+    @property
+    def is_writable(self):
+        if self.__is_writable == NULL:
+            if not self.is_environment():
+                is_writable = None
+            else:
+                is_writable = file_path_is_writable(self._magic_file)
+            self.__is_writable = is_writable
+        return self.__is_writable
+
+    def assert_exists(self):
```

Comment:
I'm not super excited about the names for these three methods. Asserts have a pretty specific meaning in Python, and I wouldn't want to confuse the caller by having them think that they are getting some sort of sub-class of an assertion error here or that the method itself is running a series of `assert` statements.

How about we take a lead from the beloved `requests` API, specifically the `raise_for_status` method on the response? We could call these methods:

- `exists_or_raise`
- `is_environment_or_raise`
- `is_writable_or_raise`

That way it's pretty clear what's happening within the method itself and the caller can guess that they will have to handle some sort of exception.

Reply:
I do like `assert_*`. That very specific meaning is exactly what I mean: raise if the following clause is not true. The `*_or_raise` form sounds a bit too Rust-y for my taste.

My intent here was to have booleans returned by the `is_*` family, and exceptions raised by the `assert_*` family. I also thought of `check_*` for the latter, but I felt that `assert` was clearer.

I'll let other folks chime in here to resolve the "tie", but I do like `assert_*` a lot.

Reply:
I'll leave this thread open in case anyone else wants to chime in on it.

Reply:
FWIW, a couple Python function definition searches that are relevant to the discussion:

- [`assert_*`](https://github.com/search?q=%2Fdef+assert_%5CS%2B%2F+language%3APython+NOT+is%3Afork&ref=opensearch&type=code): 134k files
- [`*_or_raise`](https://github.com/search?q=%2Fdef+%5CS%2B_or_raise%2F+language%3APython+NOT+is%3Afork&ref=opensearch&type=code): 5.3k files